### PR TITLE
QAT int8 MKL-DNN transformation pass with MUL 

### PR DESCRIPTION
**Pull request by bingyanghuang**
This PR is to enable the mul op in mkldnn quantization pass in slim. It is based on PR #17819 .

### CODE REVIEW

**Comment by ghost** [comment]
This PR depend on #17819

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 1]
```diff
@@ -21,23 +21,21 @@
...
+               the QuantizationFreezePass, to float32 range weights with float32 data type
+               by using the corresponding scales. This conversion is because MKL-DNN INT8 
+               conv2d kernel now only supports float32 weights input, will do weights
+               quantization inside the conv2d kernel.
+            2. Create the new conv2d op with the converted weights and link its output
+               to fake_dequantize_abs_max op's output and set conv2d's attribute "force_fp32
+               _output" as true
+            3. Transform fake_quantize_xx op to quantize op
+            4. Remove fake_dequantize_abs_max op 
+
```
Comment (lidanqing-vv) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py]: Why do you revert the change? In previous commit you put comments before `__init__`

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 2]
```diff
@@ -105,20 +98,14 @@ def apply(self, graph):
                 self.InScale[input_name] = self._load_param(self._scope,
```
Comment (lidanqing-vv) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:100]: We think the comment before this loop in line 93 is not clear. It says only about conv2d.

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 3]
```diff
@@ -140,8 +127,8 @@ def _transform_to_conv_mkldnn(self, graph, op_node):
         weight_var_node = graph._find_node_by_name(op_node.inputs, weight_name)
 
         # Set fake_dequantize_abs_max's output as new output of conv2d
-        output_var_node = graph._find_node_by_name(
```
Comment (lidanqing-vv) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:143]: Comment on line 129 `# Set fake_dequantize_abs_max's output as new output of conv2d` should be changed too. Not only conv2d is concerned.

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 4]
```diff
@@ -140,8 +127,8 @@ def _transform_to_conv_mkldnn(self, graph, op_node):
         weight_var_node = graph._find_node_by_name(op_node.inputs, weight_name)
```
Comment (lidanqing-vv) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:129]: Is `127` in line 122 supposed to be `self.s8_max`?

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 5]
```diff
@@ -169,6 +156,49 @@ def _transform_to_conv_mkldnn(self, graph, op_node):
         graph.link_to(conv_op_node, output_var_node)
         graph.safe_remove_nodes(op_node)
 
+    def _transform_to_mul_mkldnn(self, graph, op_node):
+        weight_name = op_node.input("Y")[0]
+        output_name = op_node.output("Out")[0]
+        # Convert int8 range weights to fp32 range weights
+        weight = self._load_param(self._scope, weight_name)
+        w_fp32 = np.divide(
+            np.multiply(weight, 127), self.max_range[output_name])
```
Comment (lidanqing-vv) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py]: is this `127` supposed to be `self.s8_max`?

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 6]
```diff
@@ -169,6 +156,49 @@ def _transform_to_conv_mkldnn(self, graph, op_node):
         graph.link_to(conv_op_node, output_var_node)
         graph.safe_remove_nodes(op_node)
 
+    def _transform_to_mul_mkldnn(self, graph, op_node):
+        weight_name = op_node.input("Y")[0]
```
Comment (lidanqing-vv) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:163]: why is input "Y" named weights here?

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py (hunk 3 above)
Comment (bingyanghuang) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:143]: This is in the function ` _transform_to_conv_mkldnn` , so it should be for conv2d, and for mul , the comment is in line172.

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py (hunk 6 above)
Comment (bingyanghuang) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:163]: Will sync with minghui to add a enforcement to make sure that Y should be weights. Because we will reorder the weights for int8 mul op.

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py (hunk 1 above)
Comment (bingyanghuang) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py]: It's my mistake. Has changed the order. Thanks.

Comment (bingyanghuang) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py]: Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py (hunk 2 above)
Comment (bingyanghuang) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:100]: Add "mul". Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py (hunk 4 above)
Comment (bingyanghuang) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py:129]: Done

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py (hunk 5 above)
Comment (bingyanghuang) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py]: Done

**Comment by bingyanghuang** [comment]
This PR could be merged after PR #17834 being merged.

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 7]
```diff
@@ -27,13 +27,13 @@ class TransformForMkldnnPass(object):
         1. Convert int8 range weights with float32 data type, which are generated by 
            the QuantizationFreezePass, to float32 range weights with float32 data type
            by using the corresponding scales. This conversion is because MKL-DNN INT8 
-           conv2d kernel now only supports float32 weights input, will do weights
-           quantization inside the conv2d kernel.
-        2. Create the new conv2d op with the converted weights and link its output
+           conv2d kernel and mul kernel now only supports float32 weights input, will
```
Comment (Sand3r-) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py]: `supports` -> `support`

Comment (Sand3r-) [reply python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py]: `will do weights quantization inside` -> `hence weights qunatization will happen inside`

File: python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py [hunk 8]
```diff
@@ -169,6 +158,50 @@ def _transform_to_conv_mkldnn(self, graph, op_node):
...
+            op_type='mul',
+            attrs=attrs,
+            inputs={'X': input_var_node,
+                    'Y': weight_var_node},
+            outputs={'Out': output_var_node})
+
+        # Based on the QAT's scales to calculate the scales of MKL-DNN INT8 mul
+        scale_in = self.s8_max / self.InScale[output_name]
+        scale_w = []
+        scale_w.append(self.max_range[output_name] / self.s8_max)
```
Comment (Sand3r-) [inline python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py]: Can be simplified to 
```
scale_w = [self.max_range[output_name] / self.s8_max]
```

File: python/paddle/fluid/contrib/slim/tests/test_quantization_mkldnn_pass.py [hunk 9]
```diff
@@ -162,13 +160,18 @@ def mkldnn_based_freeze_graph(self,
                             activation_quant_type + '_' + weight_quant_type,
                             marked_nodes)
         mkldnn_program = test_graph.to_program()
-        w_mkldnn = np.array(scope.find_var('conv2d_1.w_0').get_tensor())
+
+        # Check the transformation weights of conv2d and mul
+        conv_w_mkldnn = np.array(scope.find_var('conv2d_1.w_0').get_tensor())
+        mul_w_mkldnn = np.array(scope.find_var('fc_0.w_0').get_tensor())
```
Comment (Sand3r-) [inline python/paddle/fluid/contrib/slim/tests/test_quantization_mkldnn_pass.py:166]: Is there any other way of obtaining the mul and conv weights rather than by searching for their names directly?

File: python/paddle/fluid/contrib/slim/tests/test_quantization_mkldnn_pass.py [hunk 10]
```diff
@@ -162,13 +160,18 @@ def mkldnn_based_freeze_graph(self,
...
         # Check if weights are still integer
-        self.assertFalse(self.isinteger(np.sum(w_mkldnn)))
+        self.assertFalse(self.isinteger(np.sum(conv_w_mkldnn)))
+        self.assertFalse(self.isinteger(np.sum(mul_w_mkldnn)))
 
-        # Check if the conv2d output is rightly linked to fake_dequantize's 
+        # Check if the conv2d output and mul output are rightly linked to fake_dequantize's
```
Comment (Sand3r-) [inline python/paddle/fluid/contrib/slim/tests/test_quantization_mkldnn_pass.py]: `rightly` -> `correctly`

**Comment by bingyanghuang** [comment]
Close this PR because I did some stupid things . New PR will be created in an hour.
